framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,706 @@
+     */
+    public static void updateInferredMethodReturnType(ReturnNode retNode,
+            ClassSymbol classSymbol, MethodTree methodTree,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return; // Anonymous class => Ignore, for now.
+        String className = classSymbol.flatname.toString();
+
+        String jaifPath = getJaifPath(className);
+        AClass clazz = getAClass(className, jaifPath);
+
+        AMethod method = clazz.methods.vivify(JVMNames.getJVMMethodName(methodTree));
+        // Method return type
+        AnnotatedTypeMirror lhsATM = atf.getAnnotatedType(methodTree).getReturnType();
+        // Type of the expression returned
+        AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(retNode.getTree().getExpression());
+        updateAnnotationSetInScene(
+                method.returnType, atf, jaifPath, rhsATM, lhsATM,
+                DefaultLocation.RETURNS);
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene.
+     *   <ul>
+     *     <li>If the set of annotations (curAnnos) is empty, then the updated
+     *     set will be the set of annotations in newATM.</li>
+     *     <li>If curAnnos is not empty, the updated set will be the LUB
+     *     between curAnnos and newATM.</li>
+     *   </ul>
+     * <p>
+     * @param type ATypeElement of the Scene which will be modified.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update curAnnos.
+     * @param jaifPath used to identify a Scene.
+     * @param newATM the type containing a set of annotations to be
+     * added (or "lubbed") to the Scene.
+     */
+    private static void updateAnnotationSetInScene(ATypeElement type,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM, AnnotatedTypeMirror oldATM,
+            DefaultLocation defLoc) {
+        AnnotatedTypeMirror curAnnosATM = typeElementToATM(
+                type, atf, false, newATM.getUnderlyingType());
+        if (curAnnosATM.getAnnotations().size() == newATM.getAnnotations().size()
+                && newATM.getAnnotations().size() > 0) {
+            newATM = AnnotatedTypes.leastUpperBound(
+                    atf.getProcessingEnv(), atf, newATM, curAnnosATM);
+        }
+        updateTypeElementFromATM(newATM, oldATM, atf, type, 1, defLoc);
+        modifiedScenes.add(jaifPath);
+    }
+
+    /**
+     * Returns the AClass in an AScene, given a className and a jaifPath.
+     */
+    private static AClass getAClass(String className, String jaifPath) {
+        // Possibly reads .jaif file to obtain a Scene.
+        AScene scene = getScene(jaifPath);
+        return scene.classes.vivify(className);
+    }
+
+    /**
+     * Returns true if am should not be inserted in source code. This happens
+     * when am is either an implementation detail or is the default for the
```

Comment:
Rather than "implementation detail", be a bit more specific and say that it cannot be written in source code.
(Hmm, that's not true for `@InvisibleQualifier`, but it isn't necessarily an implementation detail either.)